### PR TITLE
Add user search functionality

### PR DESCRIPTION
**Pull request by Archana-PandeyM**
Added a private method _get_username_from_email(). This method will accept an email as argument and returns the username associated to provided emaiID.
This method is then being utilized by add_watcher and remove_watcher function to get the username of watcher(in case email is passed as an argument).

### CODE REVIEW

**Comment by dmranck** [comment]
Adding @advaitgupte04 as a reviewer

File: ticketutil/jira.py [hunk 1]
```diff
@@ -381,15 +381,19 @@ def remove_watcher(self, watcher):
             logger.error(error_message)
             return self.request_result._replace(status='Failure', error_message=error_message)
 
-        # If an email address was passed in for watcher param, extract the 'name' piece.
+            # If an email address was passed in for watcher param, call the _get_username_from_email method to get
+            # username for watcher .
```
Comment (dmranck) [inline ticketutil/jira.py:385]: Back up the indentation one tab level to be in-line with line 386.

File: ticketutil/jira.py [hunk 2]
```diff
@@ -514,6 +521,38 @@ def _get_watchers_list(self):
 
         return watchers_list
 
+    def _get_username_from_email(self, email):
+        """
+            Gets the username from JIRA account
+            :param email: take email id of user as an argument
+            :return: username
```
Comment (dmranck) [inline ticketutil/jira.py:528]: Indentation level should be in-line with line 525.

File: ticketutil/jira.py [hunk 3]
```diff
@@ -514,6 +521,38 @@ def _get_watchers_list(self):
...
+        """
+        rest_url_user = '{0}/rest/api/2/user'.format(self.url)
+        try:
+            r = self.s.get("{0}/search?username={1}".format(rest_url_user, email))
+            logger.debug("Get ticket content: status code: {0}".format(r.status_code))
```
Comment (dmranck) [inline ticketutil/jira.py:533]: "Get ticket content" should probably be changed to "Get username"

File: ticketutil/jira.py [hunk 4]
```diff
@@ -514,6 +521,38 @@ def _get_watchers_list(self):
...
+            r.raise_for_status()  # currently if the requested user is not found then
+            # instead of 404 API returns 200 status code and [] empty response. Reference -
+            # https://docs.atlassian.com/software/jira/docs/api/REST/7.6.1/#api/2/user-findUsers
+            user_details = r.json()
+            if user_details:
+                username = user_details[0]['self'].split("=")[1].strip()
+            else:
+                username = None
+                # workaround to handle "User not found issue" when JIRA API does not return expected error code
+                logger.error("user with email {0} could not be found".format(email))
```
Comment (dmranck) [inline ticketutil/jira.py:543]: Change "user" to "User", just because all other logging messages start with a capital letter

**Review by dmranck — CHANGES_REQUESTED** [review]
@Archana-PandeyM , this looks great! I ran through an example with the three types of Jira usernames we had discussed and they all worked for adding and removing watchers. I also tried a fake Jira email address and the error handling in _get_username_from_email handled it well, spitting out appropriate error messages, but continuing on with the execution.

I added a few small comments about indentation level and logging messages, but other than that, I would be ready to merge this soon. Thanks for working on this!

**Review by dmranck — APPROVED** [review]

**Comment by advaitgupte04** [comment]
@dmranck Archana's PR looks good to me. Tested the below scenarios: 
1. Added myself to ticket as watcher 
2. Added Andrew's username to ticket as watcher 
3. Added Jan's username to ticket as watcher
4. Tried adding a wrong username as watcher and got appropriate error

**Pull request merged**